composer/trainer/mosaic_fsdp.py

```diff
@@ -0,0 +1,210 @@
+from torch.distributed.fsdp._utils import _contains_batchnorm, _override_batchnorm_mixed_precision
+from torch.distributed.fsdp.wrap import _or_policy, _wrap, _wrap_batchnorm_individually
+
+from composer.utils import dist
+
+sharding_map = {
+    'NO_SHARD': ShardingStrategy.NO_SHARD,
+    'SHARD_GRAD_OP': ShardingStrategy.SHARD_GRAD_OP,
+    'FULL_SHARD': ShardingStrategy.FULL_SHARD,
+}
+
+
+def _get_process_group(pg, process_group_cache):
+    warnings.warn(f'Instantiating FSDP with custom process groups is an experimental feature.')
+
+    # Return regular process_groups as is, no cacheing
+    if pg is None or isinstance(pg, ProcessGroup):
+        return pg
+
+    warnings.warn(f'Composer instantiating process groups is an experimental feature.')
+
+    # Look for existing key in cache
+    if isinstance(pg, (list, tuple)):
+        pg_key = tuple(pg)
+    else:
+        pg_key = pg
+    if pg_key in process_group_cache:
+        warnings.warn(
+            f'On rank={dist.get_global_rank()} using cached progress group with {pg_key=}. ' +\
+            'Instantiate new process group if this is what was intended.'
+        )
+        return process_group_cache[pg_key]
+
+    # Handle str or List[int] process_group cases
+    if pg == 'self':
+        ranks = (dist.get_global_rank(),)
+    elif pg == 'node':
+        node_rank = dist.get_node_rank()
+        local_world_size = dist.get_local_world_size()
+        ranks = tuple(range(node_rank * local_world_size, (node_rank + 1) * local_world_size))
+    elif pg == 'local_rank_across_nodes':
+        local_rank = dist.get_local_rank()
+        local_world_size = dist.get_local_world_size()
+        num_nodes = dist.get_world_size() // dist.get_local_world_size()
+        ranks = tuple(local_rank + local_world_size * n for n in range(num_nodes))
+    elif isinstance(pg, str) and pg.startswith('set'):
+        k = int(pg.strip('set'))
+        world_size = dist.get_world_size()
+        if world_size % k:
+            raise RuntimeError(f'{world_size} must be divisible by set size ({k})')
+        start = dist.get_global_rank() // k * k
+        ranks = tuple(range(start, start + k))
+    elif isinstance(pg, str) and pg.startswith('mod'):
+        k = int(pg.strip('mod'))
+        world_size = dist.get_world_size()
+        if world_size % k:
+            raise RuntimeError(f'{world_size} must be divisible by mod ({k})')
+        ranks = tuple(range(dist.get_global_rank() % k, world_size, k))
+    elif isinstance(pg, (list, tuple)):
+        ranks = tuple(pg)
+    else:
+        raise ValueError(f'Unsure how to setup process_group={pg}')
+
+    ranks_per_subgroup_list = list(set(dist.all_gather_object(ranks)))
+    current_group, _subgroups = distributed.distributed_c10d.new_subgroups_by_enumeration(ranks_per_subgroup_list)
+    process_group_cache[pg_key] = current_group
+    return current_group
+
+
+def _pro_recursive_wrap(module: nn.Module,
```

Comment:
It's immediately not clear the pro is shorted version of process. Consider changing name to custom_recursive_wrap?